File: sdk/python/examples/07_camera_basic.py
"""摄像头基础示例：实时预览 + 拍照。

用法:
    # 在机器狗上运行 (USB 摄像头)
    python 07_camera_basic.py

    # 指定远程 RTSP 流
    python 07_camera_basic.py rtsp://192.168.66.190:8554/cam

操作:
    q — 退出
    s — 保存照片到当前目录
"""

import sys
import time

from brainstem_sdk.camera import OrixCamera

# ── 解析摄像头源 ────────────────────────────────────────────

source = 0  # 默认: USB 摄像头 index 0
if len(sys.argv) > 1:
    arg = sys.argv[1]
    # 如果是纯数字，当作设备索引
    source = int(arg) if arg.isdigit() else arg

# ── 主循环 ──────────────────────────────────────────────────

try:
    import cv2
except ImportError:
    print("需要安装 opencv-python:")
    print("  pip install brainstem-sdk[camera]")
    sys.exit(1)

photo_count = 0

with OrixCamera(source=source) as cam:
    w, h = cam.get_frame_size()
    print(f"摄像头已打开: {w}x{h}")
    print("按 'q' 退出, 按 's' 拍照")

    while True:
        frame = cam.read()
        if frame is None:
            print("读取帧失败，等待重试...")
            time.sleep(0.1)
            continue

        cv2.imshow("ORIX Camera", frame)

        key = cv2.waitKey(1) & 0xFF
        if key == ord("q"):
            break
        elif key == ord("s"):
            photo_count += 1
            filename = f"orix_photo_{photo_count:03d}.jpg"
            cam.save_photo(filename)
            print(f"已保存: {filename}")

cv2.destroyAllWindows()
print(f"共拍摄 {photo_count} 张照片。")

File: sdk/cpp/examples/07_camera_basic.cpp
// 07_camera_basic.cpp — 摄像头基本操作：打开、读帧、拍照
//
// 编译需要 OpenCV。在 CMakeLists.txt 中启用：
//   find_package(OpenCV REQUIRED)
//   target_link_libraries(07_camera_basic PRIVATE orix_sdk orix_camera ${OpenCV_LIBS})
#include "orix/orix_camera.h"

#include <iostream>
#include <thread>
#include <chrono>

int main(int argc, char* argv[]) {
    // 支持命令行指定摄像头源（索引或 URL）
    // 用法：
    //   ./07_camera_basic          # 默认 USB 摄像头 0
    //   ./07_camera_basic 1        # USB 摄像头 1
    //   ./07_camera_basic rtsp://192.168.66.190:8554/cam

    try {
        std::unique_ptr<orix::OrixCamera> cam;

        if (argc >= 2) {
            std::string src(argv[1]);
            // 如果是纯数字则作为设备索引
            bool is_number = !src.empty() && std::all_of(
                src.begin(), src.end(), ::isdigit);
            if (is_number)
                cam = std::make_unique<orix::OrixCamera>(std::stoi(src));
            else
                cam = std::make_unique<orix::OrixCamera>(src);
        } else {
            cam = std::make_unique<orix::OrixCamera>(0, 640, 480, 30);
        }

        cam->open();

        auto [w, h] = cam->get_frame_size();
        std::cout << "摄像头已打开：" << w << "x" << h << "\n";

        // ── 读取 10 帧，统计成功率 ────────────────────────────
        std::cout << "读取 10 帧...\n";
        int ok = 0;
        for (int i = 0; i < 10; ++i) {
            cv::Mat frame = cam->read();
            if (!frame.empty()) {
                ++ok;
                std::cout << "  帧 " << i + 1 << ": "
                          << frame.cols << "x" << frame.rows
                          << " channels=" << frame.channels() << "\n";
            } else {
                std::cout << "  帧 " << i + 1 << ": 读取失败\n";
            }
            std::this_thread::sleep_for(std::chrono::milliseconds(33));
        }
        std::cout << "成功率: " << ok << "/10\n\n";

        // ── 拍照保存 ──────────────────────────────────────────
        std::cout << "拍照中...\n";
        if (cam->save_photo("orix_photo.jpg")) {
            std::cout << "照片已保存: orix_photo.jpg\n";
        } else {
            std::cerr << "拍照失败。\n";
        }

        cam->close();

    } catch (const orix::CameraOpenError& e) {
        std::cerr << "摄像头错误: " << e.what() << "\n";
        return 1;
    } catch (const std::exception& e) {
        std::cerr << "错误: " << e.what() << "\n";
        return 1;
    }

    return 0;
}
